[MXNET-139] Tutorial for mixed precision training with float16

rahul003 opened the pull request:
## Description ##
Adds a FAQ page on mixed precision training with float16. Explains usage for both Gluon and Symbolic, and discusses tips to improve performance and accuracy when using mixed precision. 

https://issues.apache.org/jira/browse/MXNET-139

## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Adds tutorial
- [x] Updates symbolic examples for training and fine-tuning
- [x] Updates Gluon example

eric-haibin-lin commented on docs/tutorials/python/float16.md:
```diff
@@ -0,0 +1,280 @@
+2. Cast the data to float16 to match the input type expected by the blocks if necessary.
+
+### Training
+Let us look at an example of training a Resnet50 model with the Caltech101 dataset with float16. 
```
Add a reference link to the dataset description?

eric-haibin-lin commented on docs/tutorials/python/float16.md:
````diff
@@ -0,0 +1,280 @@
+from mxnet.gluon.data.vision.datasets import ImageFolderDataset
+```
+
+Let us start by fetching the Caltech101 dataset and extracting it. 
````
Could you add a reminder of how big the dataset is (num images, number of GBs)

eric-haibin-lin commented on docs/tutorials/python/float16.md:
```diff
@@ -0,0 +1,280 @@
+- Volta range of Nvidia GPUs
+- Cuda 9 or higher
+- CUDNN v7 or higher
+
```
Could you start with an overview that the tutorial covers both Gluon and Symbolic APIs?

eric-haibin-lin commented on docs/tutorials/python/float16.md:
````diff
@@ -0,0 +1,280 @@
+    return net
+```
+
+It is preferable to use **multi_precision mode of optimizer** when training in float16. This mode of optimizer maintains the weights in float32 even when the training is in float16. This helps increase precision of the weights and leads to faster convergence for some networks. (Further discussion on this towards the end.)
````
Do all optimizers support this mode?

rahul003 replied:
SGD supports this natively, as in there's a special kernel for that. Other optimizers support this by making a copy in Python, which can be slightly slower.

mli commented:
I feel this tutorial is a little bit too complex to reader to follow.

1. show how to write a simple dense and conv layer to use fp16, show performance numbers
2. show an end-to-end example based on another tutorial. such as http://gluon-crash-course.mxnet.io/ (it's not a good exmaple because too simplified), mentioned to users that read it first. and this tutorial will only highlight the difference for using fp16. 
3. show performance numbers and training curves

aaronmarkham commented on docs/tutorials/python/float16.md:
```diff
@@ -0,0 +1,280 @@
+
+The computational resources required for training deep neural networks has been increasing of late because of complexity of the architectures and size of models. Mixed precision training allows us to reduces the resources required by using lower precision arithmetic. In this approach we train using 16 bit floating points (half precision) while using 32 bit floating points (single precision) for o ...
+
+The float16 data type, is a 16 bit floating point representation according to the IEEE 754 standard. It has a dynamic range where the precision can go from 0.0000000596046 (highest, for values closest to 0) to 32 (lowest, for values in the range 32768-65536). Despite the decreased precision when compared to single precision (float32), float16 computation can be much faster on supported hardware.  ...
```
no comma after type

aaronmarkham commented on docs/tutorials/python/float16.md:
```diff
@@ -0,0 +1,280 @@
+
+The Volta range of Graphics Processing Units (GPUs) from Nvidia have Tensor Cores which perform efficient float16 computation. A tensor core allows accumulation of half precision products into single or half precision outputs. For the rest of this tutorial we assume that we are working with Nvidia's Tensor Cores on a Volta GPU.
+
+In this tutorial we will walk through how one can train deep learning neural networks with mixed precision on supported hardware. We will first see how to use float16 and then some techniques on achieving good performance and accuracy.
```
I'd move this to the top as the main intro, then use `## Background` for the rest.

aaronmarkham replied:
In this tutorial you will learn how you can train...
You will first see how...

Please continue on in this pattern.

aaronmarkham commented on docs/tutorials/python/float16.md:
````diff
@@ -0,0 +1,280 @@
+test_data = gluon.data.DataLoader(dataset_test, BATCH_SIZE, shuffle=False, num_workers=NUM_WORKERS)
+```
+
+Next, we'll define softmax cross entropy as our loss, accuracy as our metric and the context on which to run our training jobs. It is set by default to gpu. Please note that float16 on CPU might not be supported for all operators, as float16 on CPU is slower than float32.
````
GPU or `gpu`?

aaronmarkham commented on docs/tutorials/python/float16.md:
```diff
@@ -0,0 +1,280 @@
+
+### Finetuning
+
+You can also finetune in float16, a model which was originally trained in float32. The section of the code which builds the network would now look as follows. We first fetch the pretrained resnet50_v2 model from model zoo. This was trained using Imagenet data, so we need to pass classes as 1000 for fetching the pretrained model. Then we create our new network for Caltech 101 by passing number of  ...
```
the [model zoo](http://mxnet.incubator.apache.org/api/python/gluon/model_zoo.html).

aaronmarkham commented on docs/tutorials/python/float16.md:
```diff
@@ -0,0 +1,280 @@
+
+There are a few examples of building such networks which can handle float16 input in [examples/image-classification/symbols/](https://github.com/apache/incubator-mxnet/tree/master/example/image-classification/symbols). Specifically you could look at the [resnet](https://github.com/apache/incubator-mxnet/blob/master/example/image-classification/symbols/resnet.py) example.
+
+An illustration of the relevant section of the code is below.
```
Try to avoid above below left and right. Use follows or as follows. Or previously. This supports different reading modes.

aaronmarkham commented on docs/tutorials/python/float16.md:
```diff
@@ -0,0 +1,280 @@
+## Things to keep in mind
+
+### For performance
+1. Nvidia Tensor core essentially perform the computation D = A * B + C, where A and B are half precision matrices, while C and D could be either half precision or full precision. The tensor cores are most efficient when dimensions of these matrices are multiples of 8. This means that Tensor Cores can not be used in all cases for fast float16 computation. When training models like Resnet50 on the ...
```
cores perform or core performs

Use consistent CapitaliZation throughout.

anirudh2290 commented:
ping @rahul003

rahul003 commented:
Thanks guys for your comments. I'll address them soon and update the PR

ThomasDelteil commented on docs/tutorials/python/float16.md:
```diff
@@ -0,0 +1,280 @@
+1. [Training with Mixed Precision User Guide](http://docs.nvidia.com/deeplearning/sdk/mixed-precision-training/index.html)
+2. [Mixed Precision Training at ICLR 2018](https://arxiv.org/pdf/1710.03740.pdf)
+3. [Mixed-Precision Training of Deep Neural Networks](https://devblogs.nvidia.com/mixed-precision-training-deep-neural-networks/)
+
```
Can you add this `<!-- INSERT SOURCE DOWNLOAD BUTTONS -->` at the end of your .md file to enable the notebook download of your tutorial? Thanks!

ThomasDelteil commented on docs/tutorials/python/float16.md:
```diff
@@ -0,0 +1,280 @@
+url = "https://s3.us-east-2.amazonaws.com/mxnet-public/101_ObjectCategories.tar.gz"
+dataset_name = "101_ObjectCategories"
+data_folder = "data"
+if not os.path.isdir(data_folder):
```
these lines are un necessary as the `mx.gluon.utils.download` will create the directory if it does not exist 😃

ThomasDelteil commented on docs/tutorials/python/float16.md:
```diff
@@ -0,0 +1,280 @@
+    pretrained_net = models.get_model(name='resnet50_v2', ctx=ctx, pretrained=True, classes=1000)
+    pretrained_net.hybridize()
+    pretrained_net.cast(dtype)
+
```
a simpler way of fine-tuning a model from the model-zoo is to use `pretrained_net.output = gluon.nn.Dense(101)` and then initializing it.

piiswrong commented:
@rahul003 ping

ThomasDelteil commented:
@rahul003 ping, do you plan to finish this tutorial? It is dearly needed.

rahul003 commented:
On second thoughts as per feedback above I changed this from a runnable tutorial style to a document focusing on the changes needed to switch to Mixed precision. I updated an example in the source to add the example I had in this tutorial, and put a command to run that in this document. 

I've setup two runs whose plots I will add tomorrow. 

I'm hesitant to share raw performance numbers as those would soon become outdated as we improve. I could mention a rough speedup factor instead. What do you guys think?

rahul003 commented:
@reviewers, please check the tutorial now. I think we can merge it and keep updating it if you have suggestions for other things to add. Even as is, it would be very useful.

rahul003 commented:
can we merge this?

xinyu-intel commented on example/image-classification/benchmark_score.py:
```diff
@@ -74,11 +75,17 @@ def score(network, dev, batch_size, num_batches):
+                    if d == mx.cpu() and dtype == 'float16':
+                        #float16 is not supported on CPU
+                        continue
+                    elif net in ['inception-bn', 'alexnet'] and dt == 'float16':
```
I guess this `dt` here doesn't work, please check again. By the way, it's necessary to add a logging here to notice user what the dtype is.

rahul003 replied:
Thanks for letting me know, fixing it here https://github.com/apache/incubator-mxnet/pull/11533/
